_posts/2023-08-29-what-in-the-ml-is-going-on-around-here.md

```diff
@@ -0,0 +1,473 @@
+      "reason": "Unrecognized token 'int': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: (byte[])\"{\n    \"settings\": {\n        \"index.knn\": true,\n        \"default_pipeline\": \"i-eat-pieces-of-nlp-pipelines-for-breakfast\"\n    },\n    \"mappings\": {\n        \"properties\": {\n            \"passage_embedding\": {\n                \"type\": \"knn_vector\",\n                \"dimension\": int,\n                \"method\": {\n                    \"name\": \"string\",\n                    \"space_type\": \"string\",\n                    \"engine\": \"string\",\n                    \"parameters\": json_object\n                }\n         \"[truncated 113 bytes]; line: 10, column: 34]"
+    }
+  },
+  "status": 400
+}
+```
+
+Crap again. Fortunately, I was quick to receive a response on Slack about these calls. Once again, some assembly was required. 
+
+
+## Side Quest 3: Creating a KNN index properly. 
+
+The example call was meant to have some pieces filled in. Specifically, the `dimension` value, as well as the values inside of the `method` object. Let's fill them in. 
+
+```json
+        "dimension": int,
+        "method": {
+          "name": "string",
+          "space_type": "string",
+          "engine": "string",
+          "parameters": json_object
+        }
+```
+
+According to the [pre-trained models](https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/) page, my model has a dimensionality of 384. I'll replace `int` with `384`. 
+
+Now for the KNN `method`, I had a really hard time understanding what to put here. The only **required** field is the name. I'll take the defaults for the sake of simplicity in my examples. I'll try to boil down the docs into something a bit easier to skim. 
+
+* The engine is one of three. `nmslib`, `faiss` and `lucene`.
+* The `nmslib` engine supports `hnsw`, which is the default if you supply a name but no engine. 
+* The `faiss` engine supports the methods `hnsw` and `ivf`. 
+* The `lucene` engine supports the method `hnsw`.
+
+I'm going to just specify a method name of `hnsw` which will provide the default engine of `nmslib`. Here's the call to create my KNN index now.
+```json
+PUT /super-awesome-nlp-index
+{
+  "settings": {
+    "index.knn": true,
+    "default_pipeline": "i-eat-pieces-of-nlp-pipelines-for-breakfast"
+  },
+  "mappings": {
+    "properties": {
+      "passage_embedding": {
+        "type": "knn_vector",
+        "dimension": 384,
+        "method": {
+          "name": "hnsw"
+        }
+      },
+      "passage_text": {
+        "type": "text"
+      }
+    }
+  }
+}
+
+# The Response
+{
+    "acknowledged": true,
+    "shards_acknowledged": true,
+    "index": "super-awesome-nlp-index"
+}
+```
+
+Double rad! I made a KNN index attached to my neural search pipeline! NOW where was I? Oh right! I just want to ingest and vectorize some text, and that's it. I feel like I'm getting close. 
```

Comment:
"rad" is another slang term that will likely not be understood by all readers (and, further, is slang that particularly evokes California in the 90s, haha). I know removing these types of words can feel like your voice is being watered down, but because our goal is to be understood by as many global readers as possible, the challenge is to find words that will be broadly understood but still convey the intended meaning.